llama_index/query_engine/retry_query_engine.py

```diff
@@ -0,0 +1,292 @@
+import logging
+from typing import Any, Dict, List, Optional, Sequence
+
+from llama_index.callbacks.base import CallbackManager
+from llama_index.callbacks.schema import CBEventType
+from llama_index.data_structs.node import NodeWithScore
+from llama_index.evaluation.base import QueryResponseEvaluator
+from llama_index.indices.base_retriever import BaseRetriever
+from llama_index.indices.list.base import GPTListIndex
+from llama_index.indices.postprocessor.types import BaseNodePostprocessor
+from llama_index.indices.query.base import BaseQueryEngine
+from llama_index.indices.query.response_synthesis import ResponseSynthesizer
+from llama_index.indices.query.schema import QueryBundle
+from llama_index.indices.response.type import ResponseMode
+from llama_index.indices.service_context import ServiceContext
+from llama_index.optimization.optimizer import (
+    BaseTokenUsageOptimizer,
+    SentenceEmbeddingOptimizer,
+)
+from llama_index.prompts.prompts import (
+    QuestionAnswerPrompt,
+    RefinePrompt,
+    SimpleInputPrompt,
+)
+from llama_index.readers.schema.base import Document
+from llama_index.response.schema import RESPONSE_TYPE, Response, StreamingResponse
+
+logger = logging.getLogger(__name__)
+
+
+class RetryQueryEngine(BaseQueryEngine):
+    """Retriever query engine with retry.
+
+    Args:
+        retriever (BaseRetriever): A retriever object.
+        response_synthesizer (Optional[ResponseSynthesizer]): A ResponseSynthesizer
+            object.
+
+    """
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
```

Comment:
At a high-level, why are we rewriting logic that's in the RetrieverQueryEngine (by passing in service context / response synthesizer / retriever), instead of just passing in a BaseQueryEngine? Then the RetryQueryEngine only needs to focus on adding a retry mechanism.

I get that would be different than your design, but I think that design would be a lot simpler. Your design includes specifying the qa template and the refine prompt, and I'd love to understand justification for that decision a bit more. At a high-level I'd prefer to create more modular, simple components before creating more complex ones; the current design feels a bit hardcoded and hard to extend in the future